### ISSUE TITLE
clangd strange behaviour

### ISSUE THREAD

**Issue by nickenchev** [joaotavora/eglot#167]
Emacs: 26.1
Eglot: eglot-20181121.1924

Decided to try out clangd which was fairly easy to setup:
```
;; eglot
(require 'eglot)
(add-to-list 'eglot-server-programs '((c++-mode c-mode) "clangd"))
(add-hook 'c-mode-hook 'eglot-ensure)
(add-hook 'c++-mode-hook 'eglot-ensure)
```

I'm seeing weird behaviour when completing statements, here is an example:
I start typing my statement and get presented with my completions.
<img width="562" alt="screen shot 2018-11-22 at 3 52 28 pm" src="https://user-images.githubusercontent.com/8262271/48921658-3ef7ae00-ee6f-11e8-800d-c236d4e3e31b.png">

Then when I select the appropriate completion, I get left with the following:
<img width="193" alt="screen shot 2018-11-22 at 3 57 12 pm" src="https://user-images.githubusercontent.com/8262271/48921664-53d44180-ee6f-11e8-9a1c-a1954976fa64.png">

**Comment by mkcms** [comment]
It's a snippet. The snippets use yasnippet engine, you probably have `yas-minor-mode` enabled in your buffer. Either disable `yas-minor-mode`, or evaluate
```
(fset #'eglot--snippet-expansion-fn #'ignore)
``` 
then restart eglot.

**Comment by nickenchev** [comment]
Neither cquery nor ccls have this behaviour. Excuse my ignorance around what produces the snippets themselves, but I imagine its not clangd producing them. Why would clangd be the only server to cause this?

**Comment by nickenchev** [comment]
I use snippets all the time btw, and aren't they required for doing things like completing a function with placeholders for its parameter list?

**Comment by nickenchev** [comment]
Here is the exact same scenario, using the ccls server:
<img width="677" alt="screen shot 2018-11-22 at 4 09 11 pm" src="https://user-images.githubusercontent.com/8262271/48921903-05c03d80-ee71-11e8-878b-52668e36da32.png">
<img width="156" alt="screen shot 2018-11-22 at 4 09 23 pm" src="https://user-images.githubusercontent.com/8262271/48921907-0953c480-ee71-11e8-8fc7-2f624f70c887.png">

Still seeing (snippet) there, but no issue when pressing enter.

**Comment by mkcms** [comment]
> Neither cquery nor ccls have this behaviour.

That's because they're an optional feature of LSP. Some servers provide them, some don't.

>  Excuse my ignorance around what produces the snippets themselves, but I imagine its not clangd producing them.

The server produces them. Eglot doesn't do any server-specific things, except where it's required for the server to work as expected.

> I use snippets all the time btw, and aren't they required for doing things like completing a function with placeholders for its parameter list?

They can be used for completing functions, and indeed that's what they do in your case.

**Comment by mkcms** [comment]
> Still seeing (snippet) there, but no issue when pressing enter.

Did you restart Eglot? By that I mean, do `M-x eglot` again.

**Comment by nickenchev** [comment]
I restart emacs completely between tests.

**Comment by mkcms** [comment]
> I restart emacs completely between tests.

In that case, you need to put this:

```
(fset #'eglot--snippet-expansion-fn #'ignore)
```

in your init file.

**Comment by mkcms** [comment]
> I'm assuming this disables the snippets all together,

This line tells eglot not to expand snippets after completion. If you use yasnippet, you can still manually expand snippets as usual.

> despite the fact that clangd is producing them (as you mentioned).

In this case, eglot tells clangd that it won't expand snippets (because _you_ disabled them), so clangd won't produce snippets. And even if it did produce any snippets, eglot wouldn't use them.

> I'm not quite sure what the solution here is, or if I'm misunderstanding your suggestion.

I already gave you a solution and an explanation. The choice is up to you. You can either keep them or disable them. Or maybe I'm misunderstanding you and don't understand what is the problem you have. In this case please clarify what do you expect to happen? Thanks.

**Comment by mkcms** [comment]
@nickenchev 
Sorry, I think I closed this too quickly and didn't look carefully enough.
Is the issue that the completion should be `addChild` and it is `addChaddChild`? In that case it's a separate issue, sorry.

Can you reproduce this and post here the contents of `M-x eglot-events-buffer`? Thanks.

**Comment by nickenchev** [comment]
I'm a software developer, but my resume doesn't include "Emacs package developer", so when I say I'm misunderstanding or I'm confused, its not because I'm being a dick, its because what you said doesn't make clear sense to me.

You originally said that it's the server completely handling the snippets, and yasnippet actually expanding and handling them. So when you give me a line of code that seemingly (to my Emacs-dev untrained eye) disables snippets, and doesn't respond to my snippet expansion shortcut, you should understand my confusion. I'm not sure why you mentioned pressing TAB multiple times, it only works when there is a snippet with placeholders I can skip through, which obviously isn't the case when I have no snippets being produced.

I know you gave me an explanation, and I'm aware who's choice it is, I'm not here because I want you to tell me whether or not I should use snippets, I'm trying to understand why I'm seeing the doubled text when expanding completions.

All of the 3 servers I have tested seem to produce them, only 1 of which seems to be buggying out with.

**Comment by mkcms** [comment]
Yeah, I looked too quickly at your issue, sorry.
Can you provide a minimal program where this issue can be reproduced? Thanks.

**Comment by nickenchev** [comment]
It happens with any of my C++ projects/files I've tested with so far, when using ###clangd###. I'll write a minimal hello-world and see if it produces the issue.

**Comment by nickenchev** [comment]
Was just about to wrote a basic app to test, but even the include completion was producing the issue.

<img width="299" alt="screen shot 2018-11-22 at 4 51 01 pm" src="https://user-images.githubusercontent.com/8262271/48922759-f17f3f00-ee76-11e8-9115-6e79a2dd814c.png">

<img width="250" alt="screen shot 2018-11-22 at 4 52 01 pm" src="https://user-images.githubusercontent.com/8262271/48922762-fb08a700-ee76-11e8-94f4-bb7600036b66.png">

**Comment by mkcms** [comment]
Right, that looks like maybe something caused by `company-mode`.
Can you provide your settings for company?

**Comment by nickenchev** [comment]
Here's the output from ```eglot-events-buffer```
https://pastebin.com/B13MXmrf

**Comment by nickenchev** [comment]
```
;; Company Mode
(require 'company)
(setq company-tooltip-align-annotations t)
(setq company-dabbrev-downcase 0)
(setq company-idle-delay 0)
(with-eval-after-load 'company
  (define-key company-active-map (kbd "M-n") nil)
  (define-key company-active-map (kbd "M-p") nil)
  (define-key company-active-map (kbd "C-n") #'company-select-next)
  (define-key company-active-map (kbd "C-p") #'company-select-previous))
(delete 'company-clang company-backends)
(add-hook 'after-init-hook 'global-company-mode)
```

**Comment by joaotavora** [comment]
@nickenchev 

> I'm a software developer, but my resume doesn't include "Emacs package developer", so when I say I'm misunderstanding or I'm confused, its not because I'm being a dick, its because what you said doesn't make clear sense to me.

There's no need for this passive-aggression.  @mkcms is using his free time to help you the best he can, please cooperate.

**Comment by nickenchev** [comment]
I wasn't being passive aggressive, read the entire exchange please. We're both here spending our free time debugging an issue. I'm thankful for Eglot as it is indeed a great package, but I imagine you too are thankful for users who spend THEIR time submitting bug reports and information. My intent wasn't to be passive-aggressive, but rather simply, aggressive, after a confusing exchange that could have been very easily avoided.

**Comment by nickenchev** [comment]
> I already gave you a solution and an explanation. The choice is up to you. You can either keep them or disable them. Or maybe I'm misunderstanding you and don't understand what is the problem you have. In this case please clarify what do you expect to happen? Thanks.

I'm not sure why you'd be surprised when users respond in a negative manner after reading comments like the one above, or not have their reports read properly. I didn't take time to submit one, along with screenshots to not have it actually read. Anyways, @mkcms already noted that they hastily glanced and said sorry, we're past that and are trying to resolve the issue.

**Comment by joaotavora** [comment]
@nickenchev I did read the whole exchange. Your snarky comment isn't needed and nowhere did @mkcms deliberately disrespect you.  He made a mistake evaluating the circunstances and asked for clarification.  At no point in time did he imply you were being a dick.  Your original report doesn't very clearly mention that there is a duplicate addCh, so it is a perfectly understandable error.

Indeed if, as you say, your intent is to be aggressive after people make errors, then you have no place in this bug tracker.   In that case I hereby thank you for your report and wish you a very nice day.

**Comment by nickenchev** [comment]
So you don't see how the comment I quoted above can be easily read as condescending? Odd.
I'm not sure what age group you think your users are, but most adults don't like being talked to in that manner.

**Comment by joaotavora** [comment]
@nickenchev let's cool down for 24h

**Comment by MaskRay** [comment]
`clangd --version` is it smaller than 339543? If yes, clangd does not use `textEdit`. I am not sure how eglot handles `insertText` in `CompletionItem` when this is existing text `addCh`

**Comment by joaotavora** [comment]
@MaskRay thanks, that's most likely it.  that `clangd` reports InsertTextFormat=2 but still uses `insertText`.  I've tested and I get the behaviour described.

**Comment by joaotavora** [comment]
The problem is likely solved with 
```diff
diff --git a/eglot.el b/eglot.el
index f4a02ac..2c13538 100644
--- a/eglot.el
+++ b/eglot.el
@@ -1614,14 +1614,16 @@ is not active."
                (text-properties-at 0 comp)
              (let ((fn (and (eql insertTextFormat 2)
                             (eglot--snippet-expansion-fn))))
-               (when (or fn textEdit)
-                 ;; Undo the completion.  If before completion the buffer was
-                 ;; "foo.b" and now is "foo.bar", `comp' will be "bar".  We
-                 ;; want to delete only "ar" (`comp' minus the symbol whose
-                 ;; bounds we've calculated before) (github#160).
-                 (delete-region (+ (- (point) (length comp))
-                                   (if bounds (- (cdr bounds) (car bounds)) 0))
-                                (point)))
+               (if textEdit
+                   ;; Undo the completion.  If before completion the buffer was
+                   ;; "foo.b" and now is "foo.bar", `comp' will be "bar".  We
+                   ;; want to delete only "ar" (`comp' minus the symbol whose
+                   ;; bounds we've calculated before) (github#160).
+                   (delete-region (+ (- (point) (length comp))
+                                     (if bounds (- (cdr bounds) (car bounds)) 0))
+                                  (point))
+                 ;; a snippet should be inserted, but using plain `insertText'
+                 (delete-region (- (point) (length comp)) (point)))
                (cond (textEdit
                       (cl-destructuring-bind (&key range newText) textEdit
                         (pcase-let ((`(,beg . ,end) (eglot--range-region range)))
```
But I don't have clangd 7 here. @mkcms what do you think?

**Comment by joaotavora** [comment]
There's a  fix in ee7f9c6 but I'll wait for some testing first.

**Comment by joaotavora** [comment]
> When :insertText is present, it's open to interpretation what should happen.

Yes, the interpretation being in this case "do what we used to do with `insertText` snippets".
